Add Bayesian estimation

manchen07 opened the pull request:
Hi James,

I've added Bayesian estimation to `calc_BCSMD()` and added a few unit tests. The codes need a lot improvement, but I think it's better for you to take a look first, then we can discuss about further improvement. 

The Bayesian-estimation branch needs a rebase but there are some conflicts with the main branch. I'm not sure how to fix these without messing things up. Could you please handle the conflicts? Thank you so much!

Man

jepusto reviewed (changes requested):
Thanks for all your work on this! Overall it looks really good. My main suggestion is to pull out the code for doing the BC-SMD calculation from the posterior samples and make it a separate function, analogous to how we use `g_mlm()` when the model is estimated by REML. The separate function could be `g_mlm_Bayes)` and it would take arguments `mod`, `p_const`, and `r_const` (maybe also `r_const2`?) and carry out the effect size calculations. 

Pulling this stuff out into a separate function should make it easier to test and maintain.

jepusto commented on DESCRIPTION:
```diff
@@ -29,6 +29,7 @@ Imports:
     lmeInfo (>= 0.3.0)
 Depends: 
     nlme,
+    brms,
```
If we do this, we will just have to add code to `calc_BCSMD` that checks whether `brms` is installed and returns an informative error message if not (only if `Bayesian = TRUE`).

jepusto commented on R/calc_BCSMD.R:
```diff
@@ -333,131 +409,297 @@ calc_BCSMD <- function(design,
+    r_base_sd <- samples_r_sd[, 1:length(RE_base)]
+    r_base_var_mat <- r_base_sd^2 %*% diag(r_const_base_var)
+    r_base_var_sum <- apply(r_base_var_mat, 1, sum)
+    rho <- mean(as.vector(r_base_sd[,1])^2/ (as.vector(r_base_sd[,1])^2 + sigma_vec^2))
```
Is this consistent with how we calculate rho in `g_mlm()`? What about in designs with more than one level?

jepusto commented:
I just pushed a few small tweaks, so please be sure to pull the latest.

manchen07 commented:
I've moved brms under Suggests in DESCRIPTION. I've added the `g_mlm_Bayes()` function. I've also added more unit tests to test whether `calc_BCSMD()` runs for different models using Bayesian estimation. 

Basically, I'm interested in unit tests that
1. compare the effect size estimates and other results when using `calc_BCSMD()` versus fitting the model with `brm()` and then calculating effect sizes with `g_mlm_Bayes()`;
2. test for the calculations of `g_mlm_Bayes()`.

I've tried to do these for the Laski complex model. However, I found the results of `calc_BCSMD()` differ from these of `brm()` + `g_mlm_Bayes()` and I'm not sure why...(I comment out the comparison between Laski_comp and Laski_comp_brm due to the big discrepancies).

Could you please first review `g_mlm_Bayes()` and `calc_BCSMD()` and then take a look at the unit tests? Pease let me know if you need me to add more unit tests.

manchen07 commented:
I know what's going on with the unit tests. I forgot to set the thinning rate when using `brm()`. Now `calc_BCSMD()` should return the same results as `brm()` + `g_mlm_Bayes()`.

jepusto commented on R/calc_BCSMD.R:
```diff
@@ -285,179 +429,308 @@ calc_BCSMD <- function(design,
       if (design == "RMBB") {
         random <- eval(parse(text = paste0("list(case = ~ ", paste(c(session_RE_2, trt_RE_2), collapse = " + "),
                                            ", series = ~ ", paste(c(session_RE, trt_RE), collapse = " + "),")")))
+        random_Bayes <-paste0("(", paste(c(session_RE_2, trt_RE_2), collapse = " + "), "| case) + (",
```
I think this needs to be explict that series is nested within case. Otherwise if series is not a unique ID across all cases in the dataset, then it will get treated as crossed with case.

jepusto commented on R/calc_BCSMD.R:
```diff
@@ -285,179 +429,308 @@ calc_BCSMD <- function(design,
         random <- eval(parse(text = paste0("list(cluster = ~ ", paste(c(session_RE_2, trt_RE_2), collapse = " + "),
                                            ", case = ~ ", paste(c(session_RE, trt_RE), collapse = " + "),")")))
+        random_Bayes <- paste0("(", paste(c(session_RE_2, trt_RE_2), collapse = " + "), "| cluster) + (",
+                               paste(c(session_RE, trt_RE), collapse = " + "),"| case )")
```
Same issue here.

jepusto reviewed (changes requested):
I added some unit tests to check that the levels of the model get interpreted the same way for Bayes estimation and RML estimation. It looks like there are some outstanding problems here. See specific comments below. 

Could you please also add unit tests for the other two designs (RMBB and TR) to be sure that the levels issue is handled appropriately in these designs too?
